// @relevanceai/sdk/script/message/user.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UserMessage = void 0;
const task_js_1 = require("./task.js");
class UserMessage extends task_js_1.GenericMessage {
    /**
     * The message as text sent.
     *
     * @property {string} text
     */
    get text() {
        return this.message.content.text;
    }
    /**
     * Returns if the message triggered a subject.
     *
     * @returns {boolean}
     */
    isTrigger() {
        return this.message.content.is_trigger_message;
    }
    /**
     * The attachments sent with the message.
     *
     * @property {Attachment[]}
     */
    get attachments() {
        return (this.message.content.attachments?.map(({ file_name, file_url }) => ({
            fileName: file_name,
            fileUrl: file_url,
        })) ?? []);
    }
    /**
     * Returns if the message has attachments.
     *
     * @returns {boolean}
     */
    hasAttachments() {
        return this.attachments.length > 0;
    }
}
exports.UserMessage = UserMessage;
