// @relevanceai/sdk/esm/task/agent-strategy.js
import { stateToStatus } from "../agent.js";
import { Client } from "../client.js";
import { AgentErrorMessage, } from "../message/agent-error.js";
import { AgentMessage } from "../message/agent.js";
import { ToolMessage } from "../message/tool.js";
import { UserMessage } from "../message/user.js";
import { Task } from "./task.js";
export class AgentStrategy {
    static async get(id, agent, client = Client.default()) {
        const subject = new this(id, agent, client);
        return new Task(await subject.getMetadata(), subject);
    }
    id;
    agent;
    client;
    constructor(id, agent, client) {
        this.id = id;
        this.agent = agent;
        this.client = client;
    }
    get subject() {
        return this.agent;
    }
    async getMetadata() {
        const { metadata } = await this.client.fetch(`/agents/${this.agent.id}/tasks/${this.id}/metadata`);
        return {
            id: this.id,
            region: this.client.region,
            project: this.client.project,
            name: metadata.conversation.title,
            status: stateToStatus(metadata.conversation.state),
            createdAt: new Date(metadata.insert_date),
            updatedAt: new Date(metadata.update_date),
        };
    }
    async getMessages({ after = new Date(0) } = {}) {
        const url = `/agents/${this.agent.id}/tasks/${this.id}/view`;
        const res = await this.client.fetch(url, {
            method: "POST",
            body: JSON.stringify({
                page_size: 1_000, // @todo: pagination
                cursor: {
                    after: after.toISOString(),
                },
            }),
        });
        // message should be in ascending order
        return res.results.reverse().map((data) => {
            switch (data.content.type) {
                case "agent-error":
                    return new AgentErrorMessage(data);
                case "agent-message":
                    return new AgentMessage(data, this.agent);
                case "tool-run":
                    return new ToolMessage(data);
                case "user-message":
                    return new UserMessage(data);
                default:
                    throw new Error("unknown message response");
            }
        });
    }
}
