// @relevanceai/sdk/esm/task/workforce-strategy.js
import { Client } from "../client.js";
import { Task, } from "./task.js";
import { AgentErrorMessage, } from "../message/agent-error.js";
import { AgentMessage } from "../message/agent.js";
import { ToolMessage } from "../message/tool.js";
import { UserMessage } from "../message/user.js";
import { WorkforceAgentMessage, } from "../message/workforce-agent.js";
import { WorkforceAgentHandoverMessage, } from "../message/workforce-agent-handover.js";
export class WorkforceStrategy {
    static async get(id, workforce, client = Client.default()) {
        const subject = new this(id, workforce, client);
        const metadata = await subject.getMetadata();
        return new Task(metadata, subject);
    }
    static convertStatus() {
        return "not-implemented";
    }
    id;
    workforce;
    client;
    constructor(id, workforce, client) {
        this.id = id;
        this.workforce = workforce;
        this.client = client;
    }
    get subject() {
        return this.workforce;
    }
    async getMessages({ after: _after }) {
        const { results } = await this.client.fetch(`/workforce/items/${this.workforce.id}/tasks/${this.id}/messages`, {
            method: "POST",
            body: JSON.stringify({
                page_size: 1_000, // @todo: pagination
                // @todo: cursor "before" :S
            }),
        });
        return results.map((data) => {
            switch (data.content.type) {
                case "agent-error":
                    return new AgentErrorMessage(data);
                case "agent-message":
                    return new AgentMessage(data);
                case "tool-run":
                    return new ToolMessage(data);
                case "user-message":
                    return new UserMessage(data);
                case "workforce-agent-run":
                    return new WorkforceAgentMessage(data);
                case "workforce-agent-handover":
                    return new WorkforceAgentHandoverMessage(data);
                default:
                    throw new Error("unknown message response");
            }
        });
    }
    async getMetadata() {
        const { metadata } = await this.client.fetch(`/workforce/tasks/${this.id}/metadata`);
        return {
            id: this.id,
            region: this.client.region,
            project: this.client.project,
            name: metadata.title,
            status: WorkforceStrategy.convertStatus(), // @todo: not implemented
            createdAt: new Date(metadata.insert_date),
            updatedAt: new Date(metadata.update_date),
        };
    }
}
