// @relevanceai/sdk/esm/message/task.js
export class GenericMessage {
    message;
    constructor(message) {
        this.message = message;
    }
    /**
     * The task's message type.
     *
     * @property {TaskMessageType}
     */
    get type() {
        return this.message.content.type;
    }
    /**
     * The task message ID. Used for referencing the message.
     *
     * @property {string}
     */
    get id() {
        return this.message.item_id;
    }
    /**
     * The time the task was created.
     *
     * @property {Date}
     */
    get createdAt() {
        return new Date(this.message.insert_date_);
    }
    /**
     * Returns if the message was sent from the agent.
     *
     * @returns {boolean}
     */
    isAgent() {
        return this.type === "agent-message";
    }
    /**
     * Returns if the message is from a tool.
     *
     * @returns {boolean}
     */
    isTool() {
        return this.type === "tool-run";
    }
    /**
     * Returns if the message was sent from a user.
     *
     * @returns {boolean}
     */
    isUser() {
        return this.type === "user-message";
    }
    /**
     * Returns if the message was an error sent by the agent.
     *
     * @returns {boolean}
     */
    isAgentError() {
        return this.type === "agent-error";
    }
}
