// @relevanceai/sdk/script/task/workforce-strategy.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.WorkforceStrategy = void 0;
const client_js_1 = require("../client.js");
const task_js_1 = require("./task.js");
const agent_error_js_1 = require("../message/agent-error.js");
const agent_js_1 = require("../message/agent.js");
const tool_js_1 = require("../message/tool.js");
const user_js_1 = require("../message/user.js");
const workforce_agent_js_1 = require("../message/workforce-agent.js");
const workforce_agent_handover_js_1 = require("../message/workforce-agent-handover.js");
class WorkforceStrategy {
    static async get(id, workforce, client = client_js_1.Client.default()) {
        const subject = new this(id, workforce, client);
        const metadata = await subject.getMetadata();
        return new task_js_1.Task(metadata, subject);
    }
    static convertStatus() {
        return "not-implemented";
    }
    id;
    workforce;
    client;
    constructor(id, workforce, client) {
        this.id = id;
        this.workforce = workforce;
        this.client = client;
    }
    get subject() {
        return this.workforce;
    }
    async getMessages({ after: _after }) {
        const { results } = await this.client.fetch(`/workforce/items/${this.workforce.id}/tasks/${this.id}/messages`, {
            method: "POST",
            body: JSON.stringify({
                page_size: 1_000, // @todo: pagination
                // @todo: cursor "before" :S
            }),
        });
        return results.map((data) => {
            switch (data.content.type) {
                case "agent-error":
                    return new agent_error_js_1.AgentErrorMessage(data);
                case "agent-message":
                    return new agent_js_1.AgentMessage(data);
                case "tool-run":
                    return new tool_js_1.ToolMessage(data);
                case "user-message":
                    return new user_js_1.UserMessage(data);
                case "workforce-agent-run":
                    return new workforce_agent_js_1.WorkforceAgentMessage(data);
                case "workforce-agent-handover":
                    return new workforce_agent_handover_js_1.WorkforceAgentHandoverMessage(data);
                default:
                    throw new Error("unknown message response");
            }
        });
    }
    async getMetadata() {
        const { metadata } = await this.client.fetch(`/workforce/tasks/${this.id}/metadata`);
        return {
            id: this.id,
            region: this.client.region,
            project: this.client.project,
            name: metadata.title,
            status: WorkforceStrategy.convertStatus(), // @todo: not implemented
            createdAt: new Date(metadata.insert_date),
            updatedAt: new Date(metadata.update_date),
        };
    }
}
exports.WorkforceStrategy = WorkforceStrategy;
