// @relevanceai/sdk/esm/workforce.js
import { Client } from "./client.js";
import { TASK_PREFIX_DELIM } from "./const.js";
import { resetBackoffDuration } from "./task/task.js";
import { WorkforceStrategy } from "./task/workforce-strategy.js";
import { randomUUID } from "./utils.js";
export class Workforce {
    static async get(id, client = Client.default()) {
        const config = await client.fetch(`/workforce/items/${id}`);
        return new Workforce(config, client);
    }
    client;
    #config;
    constructor(config, client) {
        this.#config = config;
        this.client = client;
    }
    get id() {
        return this.#config.id;
    }
    get region() {
        return this.client.region;
    }
    get project() {
        return this.client.project;
    }
    get name() {
        return this.#config.workforce_metadata.name;
    }
    async getTask(id) {
        return await WorkforceStrategy.get(id, this, this.client);
    }
    async sendMessage(message, task) {
        let taskId;
        // embed keys require a task prefixing for new tasks
        if (!task && this.client.isEmbedKey()) {
            taskId = [this.client.key.taskPrefix, await randomUUID()].join(TASK_PREFIX_DELIM);
        }
        else if (task) {
            taskId = task.id;
        }
        const { workforce_task_id: taskIdResponse } = await this.client.fetch("/workforce/trigger", {
            method: "POST",
            body: JSON.stringify({
                workforce_id: this.id,
                workforce_task_id: taskId,
                trigger: {
                    message: {
                        role: "user",
                        content: message,
                        attachments: [],
                    },
                },
            }),
        });
        if (task) {
            task[resetBackoffDuration]();
        }
        return task ?? this.getTask(taskIdResponse);
    }
}
